scapy/contrib/bgp.py

```diff
@@ -1,169 +1,676 @@
-	name = "BGP Optional Parameters"
-	fields_desc = [
-	ByteField("type", 2),
-	ByteField("len", None),
-	StrLenField("value", "",  length_from = lambda x: x.len),
-	]
-	def post_build(self,p,pay):
-		if self.len is None:
-			l = len(p) - 2 # 2 is length without value
-			p = p[:1]+struct.pack("!B", l)+p[2:]
-		return p+pay
-	def extract_padding(self, p):
-		"""any thing after this packet is extracted is padding"""
-		return "",p
+    """The first part of any BGP packet"""
+    name = "BGP header"
+    fields_desc = [
+        XBitField("marker", 0xffffffffffffffffffffffffffffffff, 0x80),
+        ShortField("len", None),
+        ByteEnumField("type", 4, {0:"none",
+                                  1:"open",
+                                  2:"update",
+                                  3:"notification",
+                                  4:"keep_alive"}),
+    ]
+    def post_build(self, p, pay):
+        if self.len is None:
+            l = len(p)
+            if pay is not None:
+                l += len(pay)
+                p = p[:16]+struct.pack("!H", l)+p[18:]
+        return p+pay
 
-class BGPOpen(Packet):
-	""" Opens a new BGP session"""
-	name = "BGP Open Header"
-	fields_desc = [
-	ByteField("version", 4),
-	ShortField("AS", 0),
-	ShortField("hold_time", 0),
-	IPField("bgp_id","0.0.0.0"),
-	ByteField("opt_parm_len", None),
-	PacketListField("opt_parm",[], BGPOptionalParameter, length_from=lambda p:p.opt_parm_len),
-	]
-	def post_build(self, p, pay):
-		if self.opt_parm_len is None:
-			l = len(p) - 10 # 10 is regular length with no additional options
-			p = p[:9] + struct.pack("!B",l)  +p[10:]
-		return p+pay
+#
+# -------------------  BGP Open message and capability negotiation
+#
+class BGPOptionalParameter(PadPacket):
+    """Format of optional Parameter for BGP Open"""
+    name = "BGP Optional Parameters"
+    fields_desc = [
+        ByteField("type", 2),
+        FieldLenField("len", None, fmt="B", length_of="Value"),
+        StrLenField("value", "", length_from=lambda x: x.len),
+    ]
+    def post_build(self, p, pay):
+        if self.len is None:
+            l = len(p) - 2 # 2 is length without value
+            p = p[:1]+struct.pack("!B", l)+p[2:]
+        return p+pay
 
+class Capability(BGPOptionalParameter):
+    """Default capability implementation"""
+    name = "MultiProtocol address family support"
+    fields_desc = [
+        ByteField("type", 2),
+        ByteField("len", None),
+        ByteField("capa_type", 1),
+        FieldLenField("capa_len", None, fmt="B", length_of="capa_value"),
+        StrLenField("value", "", length_from=lambda x: x.capa_len),
+    ]
+    def post_build(self, p, pay):
+        if self.len is None:
+            l = len(p) - 2      # length of the packet after 'len'
+            p = p[:1]+struct.pack("!B", l)+p[2:]
+        if self.capa_len is None:
+            cl = len(p) - 4     # length of the packet after 'capa_len'
+            p = p[:3]+chr(cl)+p[4:]
```

Comment:
Is there a reason why you use `chr` here and `struct.pack` three lines above?

BTW, there also, you can replace `!B` by `B`.